### ISSUE TITLE
Configurer la Content Security Policy

### ISSUE THREAD

**Issue by sandhose** [zestedesavoir/zds-site#3242]
La [CSP](https://developer.mozilla.org/en-US/docs/Web/Security/CSP/Using_Content_Security_Policy) se configure via un Header HTTP (d'où le tag infra ; ca se fera dans la conf nginx) et permet de donner liste explicite des provenance de chaque type de resources (script, images, font, stylesheets...), afin d'éviter au maximum les failles XSS.

Il faut donc se fixer sur une liste de domaines pour chaque resource.
- `default-src`: `self`
- `child-src`: A déterminer, a priori y'a la liste des sites de vidéos supportés + JSFiddle
- `connect-src`: `self`
- `font-src`: 
  - `self`
  - `https://fonts.googleapis.com`
  - `https://fonts.gstatic.com`
- `form-action`: `self`
- `img-src`: `*`
- `media-src`: `none` - A priori on utilise pas de `<audio>` ni `<video>`
- `object-src`: `none` (à moins qu'on ait des applets Java/du Flash qui traine quelque part)
- `script-src`:
  - `self`
  - `https://cdn.mathjax.org`
  - `https://www.google-analytics.com`
  - `https://www.googletagmanager.com` (on utilise encore ce truc ?!)
- `style-src`:
  - `self`
  - `https://fonts.googleapis.com`
- `upgrade-insecure-requests`: oui, non ? le mettre empêcherait tout chargement sur des serveurs ne supportant pas l'HTTPS, mais upgraderait toute requête HTTP en HTTPS

ping @Eskimon @Situphen @SpaceFox pour votre avis + correction pour d'éventuels oublis

**Comment by SpaceFox** [comment]
> upgrade-insecure-requests: oui, non ?

Non, ça pèterait la moindre image hébergée sur un serveur ne supportant pas HTTPS.

**Comment by artragis** [comment]
juste par curiosité, les paramètres que tu montres sont les noms "côté nginx" ou ce sont des choses à mettre côté "settings_prod.py"?

**Comment by gustavi** [comment]
Pour moi c'est côté nginx.

**Comment by artragis** [comment]
je pense aussi, mais la syntaxe telle que donnée ne me fait pas penser à un
fichier nginx là. on a peu de ressource sysadmin alors vaut mieux éviter de
demander @spacefox de faire 50 000 recherches pour gérer ce genre de
paramètres.

2015-12-15 10:09 GMT+01:00 Laville Augustin notifications@github.com:

**Comment by SpaceFox** [comment]
Surtout que c'est le boulot de @Un-Cornichon :)

**Comment by sandhose** [comment]
Tout ça, c'est côté nginx. C'est un header à rajouter, qui ressemblerait à ça:

```
Content-Security-Policy: default-src 'self' ; script-src 'self' cdn.mathjax.org, www.google-analytics.com, www.googletagmanager.com; style-src 'self' fonts.googleapis.com; img-src * ; font-src 'self' fonts.googleapis.com, fonts.gstatic.com;
```

> >  upgrade-insecure-requests: oui, non ?
> 
> Non, ça pèterait la moindre image hébergée sur un serveur ne supportant pas HTTPS.

Il y a aussi dans ce header le moyen de donner une URL de report, où le navigateur va envoyer une liste de toutes les resources pas chargées en HTTPS, plus toutes les "violations" de la CSP (donc les potentielles failles XSS). Si on a pas envie/le temps de dev un backend pour ça, y'a le service (à priori gratuit) [report-uri.io](https://report-uri.io) qui récupère ces rapports, et permet d'analyser les données. 

Si on se rends compte que toutes les images au final supporteraient l'HTTPS, on pourrait éventuellement reconsidérer cette option. 

Aussi, on aurait de quoi prévenir les utilisateurs si leur site supporte pas l'HTTPS (parce que globalement, les seuls sites avec des images linké sur ZdS qui supportent pas l'HTTPS, c'est des sites perso gérés par des membres), et les inciter à songer à obtenir un certificat via Letsencrypt ou autre. 

Faut savoir que Chrome (et peut-être Firefox aussi) prévoit de marquer le site comme "non sécurisé" (le fameux cadenas rouge barré) d'ici quelques mois en cas de "mixed-content" (sachant qu'actuellement, visuellement, on "perd" juste le cadenas vert, pour avoir l'icône par défaut, comme si l'on est sur une page en HTTP)

**Comment by SpaceFox** [comment]
Par moment à vous lire, je me demande si ZdS n'est pas un site bancaire.

**Comment by artragis** [comment]
> Faut savoir que Chrome (et peut-être Firefox aussi) prévoit de marquer le site comme "non sécurisé" (le fameux cadenas rouge barré) d'ici quelques mois en cas de "mixed-content" (sachant qu'actuellement, visuellement, on "perd" juste le cadenas vert, pour avoir l'icône par défaut, comme si l'on est sur une page en HTTP)

Je rappelle que les "mixed content" ne concernent que **le domaine zestedesavoir.com** donc on s'en fiche des autres sites.

Ce qu'il nous faudrait c'est forcer les urls en "https" pour les images des galeries (donc qu'on trouve dans les tuto) ou bien mettre par défaut l'url sans le "http://zestedesavoir.com". Il serait bien plus efficace de mettre "/media/gallerie/......".
donc ç part forcer côté nginx les images à être chargées en https il n'y a rien d'autre à faire.
forcer à charger en https les images externes c'est contreproductif.

**Comment by gustavi** [comment]
Effectivement on devrait pouvoir charger les galeries comme ça, est-ce vraiment compliqué à changer ?

**Comment by sandhose** [comment]
Bah le problème, c'est surtout les messages existants ; soit on fait ça à coup de grosse requête SQL, soit on fait ça à la volée au moment du rendu markdown. La première solution est plus clean, mais sûrement plus compliquée à mettre en place.

**Comment by sandhose** [comment]
Après, pour les nouveaux messages, on peux filtrer directement au moment d'envoyer à coup de RegEx

**Comment by artragis** [comment]
bah si côté nginx on force la redirection https ça peut aller non?

sinon au moins il faut que @UnCornichon nous mette la CSP comme tu l'as
proposé.

**Comment by ghost** [comment]
J'ai un petit peu trop pris mon temps sur ce coup, la fin d'année dernière a été particulière, mais voilà c'est fait sur la bêta déjà.

**Comment by SpaceFox** [comment]
Cool, merci !

**Comment by artragis** [comment]
Et de manière étrange aucun header n'est détecté par les navigateur.

**Comment by sandhose** [comment]
J'ai corrigé le problème et ajusté le header pour qu'on ait plus d'erreur. Au final, j'ai ça comme header:

```
Content-Security-Policy: default-src 'self'; script-src 'self' 'unsafe-eval' 'unsafe-inline' https://cdn.mathjax.org https://www.google-analytics.com https://www.googletagmanager.com; style-src 'self' 'unsafe-inline' https://fonts.googleapis.com; img-src * ; font-src 'self' https://fonts.googleapis.com https://fonts.gstatic.com; connect-src 'self' ; media-src 'none' ; object-src 'none' ; child-src 'none' ;
```

**Comment by artragis** [comment]
Sur la beta c'est OK. Les headers sont là et j'ai pas vu de problèmes avec les images ou ce qui existe.
Vous pouvez donc mettre ça sur la prod quand vous avez le temps.

**Comment by sandhose** [comment]
Etant donné qu'on a encore beaucoup de soucis à cause de ça, je pense qu'il vaut mieux ne pas mettre ça en place sur la beta. J'ai donc commenté le header dans la config nginx de la bêta, et on verra ça sûrement à la prochaine release.

---

```
add_header Content-Security-Policy "default-src 'self'; script-src 'self' 'unsafe-eval' 'unsafe-inline' https://www.gstatic.com https://www.google.com https://cdn.mathjax.org https://www.google-analytics.com https://www.googletagmanager.com; style-src 'self' 'unsafe-inline' https://fonts.googleapis.com; img-src * ; font-src 'self' https://fonts.googleapis.com https://fonts.gstatic.com; connect-src 'self' ; media-src 'none' ; object-src 'none' ; child-src https://www.google.com; upgrade-insecure-requests ;";
```

**Comment by artragis** [comment]
du coup?

**Comment by Situphen** [comment]
Up ! Ça a été fait ou pas ?

**Comment by gllmc** [comment]
Pas encore, [visiblement](https://cspvalidator.org/#url=https://zestedesavoir.com) (idem sur la bêta).

**Comment by motet-a** [comment]
Vous pouvez aussi `curl -s -o /dev/null -D - https://zestedesavoir.com | grep -i content-security-policy`.

**Comment by artragis** [comment]
On en est où ici @sandhose @Situphen ?
